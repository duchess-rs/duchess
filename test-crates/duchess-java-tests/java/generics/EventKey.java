package generics;

public enum EventKey implements MapKey {
    A("abc"),
    B("cde");

    private final String name;

    private EventKey(String name) {
        this.name = name;
    }

}