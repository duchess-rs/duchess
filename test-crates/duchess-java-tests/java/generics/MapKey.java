package generics;

public interface MapKey {

}
