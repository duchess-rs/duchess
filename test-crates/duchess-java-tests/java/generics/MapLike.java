package generics;

import java.util.HashMap;

public class MapLike<T extends MapKey> {
    public MapLike() {
        this.storage = new HashMap<>();
    }

    private HashMap<T, Object> storage;

    public void add(T key, Object value) {
        storage.put(key, value);
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (T key : storage.keySet()) {
            sb.append(key + "=" + storage.get(key) + "\n");
        }
        return sb.toString();
    }
}
