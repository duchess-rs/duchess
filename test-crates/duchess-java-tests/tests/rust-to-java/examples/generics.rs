//@run
use duchess::{java, prelude::*};

duchess::java_package! {
    package generics;

    public final class generics.EventKey implements generics.MapKey {
        public static final generics.EventKey A;
        public static final generics.EventKey B;
    }

    interface generics.MapKey { * }


    public class generics.MapLike<TT extends generics.MapKey> {
        public generics.MapLike();
        public void add(TT, java.lang.Object);
        public <T extends generics.MapKey> void methodGeneric(T, java.lang.Object);
    }
}

fn main() {
    use generics::*;
    let base = EventKey::get_a();
    let event = MapLike::<EventKey>::new().execute().unwrap();
    event.add(base, "value-a").execute().unwrap();
    event
        .method_generic::<EventKey>(EventKey::get_b(), "value-b")
        .execute()
        .unwrap();
    assert_eq!(
        event.to_string().execute().unwrap(),
        Some("A=value-a\nB=value-b\n".to_string())
    );
}
