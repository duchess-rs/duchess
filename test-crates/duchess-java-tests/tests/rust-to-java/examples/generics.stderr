warning: unused import: `java`
 --> tests/ui/examples/generics.rs:2:15
  |
2 | use duchess::{java, prelude::*};
  |               ^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: 1 warning emitted

